ESBMC-Solidity

kunjsong01 opened the pull request:
Added a new Solidity frontend to enable SMT-based verification of Solidity smart contracts. 

Note: 
This new frontend relies on a third party JSON library called nlohmann json. There are a few CMake options to integrate nlohmann json library with another project: 

https://json.nlohmann.me/integration/cmake/ 

For the time being, I've used the "Embedded" option described in the link above, which copies the entire source tree in a subdirectory `src/solidity-frontend/nlohmann_json` and added `add_subdirectory(nlohmann_json)` in `src/solidity-frontend/CMakeLists.txt`. 

There is another option called "Embedded (FetchContent)", which obviates the need to copy the entire source tree of nlohmann json library, and hence makes our code base lighter. However, it would require FetchContent in CMake v3.11.

kunjsong01 committed:
Adding solidity ast frontend placeholder

Signed-off-by: Kunjian Song <kunjian90@gmail.com>

kunjsong01 committed:
Fixed linking and extension error

kunjsong01 committed:
Lifting barrier in parse() function

kunjsong01 committed:
Added code placeholders for Solidity's ast_convert, parse_treet, declarationt, symbolt, exprt and irept

kunjsong01 committed:
Fixed issues in expr, irep, symbol and added location

kunjsong01 committed:
reverted 9a6501bf0 and 95b31232

kunjsong01 committed:
added -A option for Solidity plain text AST

kunjsong01 committed:
store AST text and json contents in stream objects

kunjsong01 committed:
improved encapsulation by adding libUtil in solidity frontend

kunjsong01 committed:
ported SIF utilities, AST nodes and visitors

kunjsong01 committed:
Fixed embedded nlohmann::nlohmann CMake files

kunjsong01 committed:
embedded SIF into ESBMC

kunjsong01 committed:
added debug prints in libSif/*

kunjsong01 committed:
added context converter placeholder

kunjsong01 committed:
added internal intrinsics and started to convert and add symbols to the context

kunjsong01 committed:
__ESMC_assume: flattened the hierarchical data of clang declarations

kunjsong01 committed:
Added ESBMC_intrinsics object in json to hold all ESBMC or TACAS definitions

kunjsong01 committed:
separate intrincis and ast nodes

kunjsong01 committed:
Added solidity types converter functions and started to process DeclFunction

kunjsong01 committed:
added solidity declaration tracker

kunjsong01 committed:
added json tracker configure file

kunjsong01 committed:
Changed design of decl tracker: use homogeneous tracker instead of a unified namespace

kunjsong01 committed:
added get_type for Builtin type

kunjsong01 committed:
finished q_type parsing in get_type function

kunjsong01 committed:
completed checks for isVariadic and isInlined

kunjsong01 committed:
added get_location_from_decl and get_presumed_location in solidity_convert

kunjsong01 committed:
added set_location and finished locationt setup for DeclFunction

kunjsong01 committed:
completed get_decl_name

kunjsong01 committed:
completed first symbol instantiation

kunjsong01 committed:
added moduleName and the corresponding parsing functions

kunjsong01 committed:
using jsonTrackerRef to improve readibility

kunjsong01 committed:
Successfully added first symbol

kunjsong01 committed:
started to process function params

kunjsong01 committed:
Completed tracker's populate_params

kunjsong01 committed:
improved funcParam struct

kunjsong01 committed:
Completed function param processing

kunjsong01 committed:
Changed -A to -M to hold the dummy main for Solidity verification

kunjsong01 committed:
added member clang_c_module in solidity_ast_language

kunjsong01 committed:
Converted ESBMC intrinsics and dummy main to symbols

kunjsong01 committed:
Extracted ast nodes from json and started to convert

kunjsong01 committed:
clean up

kunjsong01 committed:
restored clang_c_converter

kunjsong01 committed:
Var _x symbol Part I
  - get_type: added get_type, get_sub_type and get_buildin_type

kunjsong01 committed:
Var _x symbol Part II:
  - Completed get_decl_name
  - TODO: Completed get_location

kunjsong01 committed:
Var _x symbol Part III
  - Completed get_decl_location
  - Completed get_default_symbol and added the symbol to the context
  - Completed adjust symbol

kunjsong01 committed:
Function symbol Part I:

  - refactored decl tracker and added inheritance
  - completed get_type
  - completed location
  - completed decl_name
  - added symbol to the context

kunjsong01 committed:
Function symbol Part II:

  - Function args: zero param done
  - BinaryOperator: BO_Assign done (TODO)

kunjsong01 committed:
Function symbol Part II - continued 1

kunjsong01 committed:
Function symbol Part II - continued 2

kunjsong01 committed:
Function symbol Part II - continued 3

  - DeclRefExprClass

kunjsong01 committed:
Function symbol Part II - continued 4

  - ImplicitCastExprClass
  - IntegerLiteralClass

kunjsong01 committed:
Function symbol Part II - continued 4:

  - completed BinaryOperatorExpr symbols for:
    * _x=100;
    * _y=240

kunjsong01 committed:
Function symbol Part III

  - completed BinaryOperatorExpr: BO_Add and BO_Assign for:
    sum = _x + _y;
  - NOTE: removed unnecessary implicit cast to simplfy the process

kunjsong01 committed:
Function symbol Part IV

kunjsong01 committed:
Function symbol Part IV - continued 1:

  - Completed CalleExpr up to args

kunjsong01 committed:
Fix LValueToRValue cast

kunjsong01 committed:
Function symbol Part IV - continued 2:

  - Set up the skeleton for BinaryOpExpr "sum > 100"

kunjsong01 committed:
Function symbol Part IV - continued 3:

  - Completed all symbol conversion

kunjsong01 committed:
Rebased branch to ESBMC latest master

kunjsong01 committed:
Fixed solidity frontend

  - fixed solidity_ast_language module
  - fixed -M argument for dummy main
  - restored Solidity frontend to check overflow
  - added the new msg handler to solidity_converter module

kunjsong01 committed:
Code clean-up Part I:

  - renamed solidity-ast-* to solidity-frontend

kunjsong01 committed:
Code clean-up Part II:

  - removed unused SIF libraries

kunjsong01 committed:
Separate pattern-based check from semantic analysis

kunjsong01 committed:
Code refactoring: overflow - Part I

  - state variable declaration
  - re-wrote conversions based on Solidity grammar rules
    to :
    >> improve the readability of the code
    >> facilitate the maintenance and extension of the code

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch1

  - function definition and body
  - re-wrote conversions based on Solidity grammar rules
    >> added rule parameter-list
    >> added rule function-definition
    >> added rule expression
    >> improved readability, maintenance and extension of the code

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch2

  - make ready for rule expression conversions

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch3

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch4

  - Completing the refactoring!

  Experiment of improvements:
  - Removed all ImplicitCastExpr of the type <IntegralCast>
  - Removed all ImplicitCastExpr of the type <LValueToRValue>
  - CallExpr uses hard-coded "bool" instead of the oiginal "signed_int"
    This is set based on Solidity AST node.

  All experiment passed!

kunjsong01 committed:
Added BO_Sub and BO_LT to verify underflow

kunjsong01 committed:
Stop execution after detecting Tx.Origin vulnerability

kunjsong01 committed:
Fixed the code to use typeDescription instead of typeName

kunjsong01 committed:
Fixed the consistency problem of typeName vs. typeDescriptions

kunjsong01 committed:
Added support for nondet function - Part I

kunjsong01 committed:
Added support for nondet function - Part II

kunjsong01 committed:
Fixed symbol table printing

Signed-off-by: kunjsong01 <kunjian90@gmail.com>

kunjsong01 committed:
Added test case for PR #485 - Fixed symbol table printing

Signed-off-by: kunjsong01 <kunjian90@gmail.com>

kunjsong01 committed:
Fixed function side effect

kunjsong01 committed:
Fixed function return

kunjsong01 committed:
Added support for nondet function - Part III

kunjsong01 committed:
Added support for nondet function - Part IV

kunjsong01 committed:
Added support for nondet function - Part V

kunjsong01 committed:
Added support for init value of local var declarations

kunjsong01 committed:
Added support for assume function and binary operator != BO_NE

kunjsong01 committed:
Added support for modulo operator %

kunjsong01 committed:
Added support for __VERIFIER_assume

kunjsong01 committed:
Added support for UO_PreDec

kunjsong01 committed:
Added support for ForStatement

kunjsong01 committed:
Fixed ForStmt

kunjsong01 committed:
Added support for array verfication - Part I

  - Fixed duplicate current function name
  - Added support to convert function parameters

kunjsong01 committed:
Added support for array verfication - Part II

kunjsong01 committed:
Added support for array verfication - Part III

  - added code block to handle ArrayTypeName

kunjsong01 committed:
Added support for array verfication - Part IV

  - Added IndexAccess (ArraySubscriptExprClass)
    and ArrayToPointer decay
  - Completed "else" block in if-then-else

kunjsong01 committed:
Code clean-up Part I:

  - removed unused files and updated CMake

kunjsong01 committed:
Merge branch 'master' of https://github.com/kunjsong01/esbmc

kunjsong01 committed:
Removed debug lines

mikhailramalho commented:
Before I start to review this: I saw you pushed the whole https://github.com/nlohmann/json repo here. Are you using it?

lucasccordeiro commented:
Shall we add https://github.com/nlohmann/json to our building instructions instead of copying this code to our repository?

feliperodri commented:
> Shall we add https://github.com/nlohmann/json to our building instructions instead of copying this code to our repository?

Either that or add it as a [sub-module](https://git-scm.com/book/en/v2/Git-Tools-Submodules) instead of copying it for sure.

kunjsong01 commented:
```
I saw you pushed the whole https://github.com/nlohmann/json repo here. Are you using it?
```
Yes, the nlohmann json library is used to process Solidity JSON AST. 

```
Shall we add https://github.com/nlohmann/json to our building instructions instead of copying this code to our repository?
```
That's what I thought too. We have two options: 

1. Use CMake's FetchContent (available in CMake 3.11+)
2. Use Git submodule

lucasccordeiro commented:
Git submodule seems to be a good option.

mikhailramalho commented:
Git submodule usually causes a bunch of annoyances.

Can we add it as a dependency to esbmc, like boost, the solvers, etc.? Is
it easily available to download, e.g, via apt, vcpkg, conan, etc.?

Em seg., 13 de set. de 2021 às 11:19, Lucas Cordeiro <
***@***.***> escreveu:

> Git submodule seems to be a good option.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/493#issuecomment-918300867>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZE4IVKKD7E2JFOCL3UBYI67ANCNFSM5D574E7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

kunjsong01 commented:
> Git submodule usually causes a bunch of annoyances. Can we add it as a dependency to esbmc, like boost, the solvers, etc.? Is it easily available to download, e.g, via apt, vcpkg, conan, etc.? 

According to this [info page](https://json.nlohmann.me/integration/package_managers/), it can be downloaded from popular package managers such as Homebrew, vcpkg, Conan, cget etc.

mikhailramalho commented:
It also seem to be available via apt:

$ sudo apt search nlohmann
Sorting... Done
Full Text Search... Done
nlohmann-json3-dev/focal,focal 3.7.3-1 all
  JSON for Modern C++

Is v3.7.3 enough for you? I'm using Ubuntu 20.04 LTS.

If that's enough, we can simply add the instruction to the readme (like
boost) and link the library.

Also, please remove the commits that add/change the nlohmann-json repo code
from the PR.



Em seg., 13 de set. de 2021 às 11:37, KSong ***@***.***>
escreveu:

> Git submodule usually causes a bunch of annoyances. Can we add it as a
> dependency to esbmc, like boost, the solvers, etc.? Is it easily available
> to download, e.g, via apt, vcpkg, conan, etc.?
>
> According to this info page
> <https://json.nlohmann.me/integration/package_managers/>, it can be
> downloaded from popular package managers such as Homebrew, vcpkg, Conan,
> cget etc.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/493#issuecomment-918318077>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHYXYLWOXB5RXG3MJ73UBYLELANCNFSM5D574E7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

rafaelsamenezes commented:
> According to this info page, it can be downloaded from popular package managers such as Homebrew, vcpkg, Conan, cget etc.

We could setup the CI to use vcpkg however,

The latest version of CMake on Ubuntu 21.04 already has the fetch content module. So for me, the best option is to use it since it will work for every OS and architecture.

mikhailramalho commented:
I still think that we should at least support the latest LTS version, which
is 20.04.

That being said, Ubuntu 20.04 ships with cmake 3.16.3, so it should be safe
to use vcpkg.

(should we move all our dependencies fetching to vcpkg?)

Em seg., 13 de set. de 2021 às 11:47, Rafael Sá Menezes <
***@***.***> escreveu:

> According to this info page, it can be downloaded from popular package
> managers such as Homebrew, vcpkg, Conan, cget etc.
>
> We could setup the CI to use vcpkg however,
>
> The latest version of CMake on Ubuntu 21.04 already has the fetch content
> module. So for me, the best option is to use it since it will work for
> every OS and architecture.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/493#issuecomment-918327070>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6KPF6HCVMKUBKYK33UBYMIBANCNFSM5D574E7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

rafaelsamenezes commented:
> still think that we should at least support the latest LTS version, which is 20.04

The current CI scripts are targeting for Ubuntu latest. We have to fix it then. 

> That being said, Ubuntu 20.04 ships with cmake 3.16.3

Fetch content requires [3.11](https://cmake.org/cmake/help/latest/module/FetchContent.html)

So this should be easiest.

> should we move all our dependencies fetching to vcpkg?

Most of the environment already works through vcpkg (at least in windows). The exceptions are the solvers since only Z3 is available in vcpkg

mikhailramalho commented:
> Most of the environment already works through vcpkg (at least in windows).
> The exceptions are the solvers since only Z3 is available in vcpkg
>
>
>
But can't you also use cmake's fetch thing to build a dependency? I thought
I read this somewhere.

@kunjsong01, this discussion is not related to your PR, just the part to
add the support for the json library via vcpkg.

-- 

Mikhail Ramalho.

rafaelsamenezes commented on src/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ add_subdirectory (cpp)
 endif()
 
 add_subdirectory (clang-c-frontend)
+add_subdirectory (solidity-frontend)
```
Could you add this as an option? https://github.com/esbmc/esbmc/blob/master/scripts/cmake/Options.cmake#L24

kunjsong01 replied:
ACK

kunjsong01 replied:
Done

rafaelsamenezes commented on src/solidity-frontend/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+# replace it with FetchContent of CMake v3.11
+# to automatically download nlohmann_json as a dependency at configure stage
+add_subdirectory(nlohmann_json)
```
Move this dependency to here: https://github.com/esbmc/esbmc/blob/master/scripts/cmake/ExternalDependencies.cmake

kunjsong01 replied:
ACK

kunjsong01 replied:
Done

rafaelsamenezes commented on src/solidity-frontend/solidity_language.h:
```diff
@@ -0,0 +1,75 @@
+#include <util/c_link.h> // for c_link
+#include <util/language.h>
+#include <fstream>
+#include <iostream>
```
Avoid using iostream if you can

kunjsong01 replied:
ACK

kunjsong01 replied:
Done

rafaelsamenezes commented on src/util/cmdline.cpp:
```diff
@@ -121,6 +123,14 @@ bool cmdlinet::parse(
   }
+  if (vm.count("dummymain")) // dummy main for Solidity frontend
+  {
+    sol_main = vm["dummymain"].as<std::vector<std::string>>().back();
```
Could you explain why "--function" is not enough?

kunjsong01 replied:
The new Solidity frontend relies on clang_c_frontend to convert ESBMC intrinsics (Line 34 and Line 77 in `src/solidity-frontend/solidity_language.cpp`):

```cpp
34   clang_c_module->parse(sol_main_path, msg);
...
77   clang_c_module->convert_intrinsics(new_context, msg);
```

If only '--function' is provided but 'dummymain' is not provided, this conversion will fail with the error message below: 
```cpp
error: no input files
Assertion failed: (Jobs.size() == 1), function buildASTs, file ../src/clang-c-frontend/AST/build_ast.cpp, line 90.
```

'dummymain' was added as a workaround to resolve this error.

rafaelsamenezes commented on src/util/cmdline.cpp:
```diff
@@ -121,6 +123,14 @@ bool cmdlinet::parse(
+  }
+  if (vm.count("function")) // We need Solidity function name for pattern-based verification
+  {
+    sol_function = vm["function"].as<std::string>();
```
Could you replace with another name so it will not conflict with the "--function" flag?

kunjsong01 replied:
Sorry, I think my comment in line 130 appears confusing. Please ignore it.

The intention was to re-use the "function" flag defined in `src/esbmc/options.cpp`: 

```cpp
110    {{"function",
111      boost::program_options::value<std::string>()->value_name("name"),
112      "set main function name"},
```

If we create a new flag, would that be a duplicate?

rafaelsamenezes replied:
> The intention was to re-use the "function" flag defined in src/esbmc/options.cpp:

Are you sure this is needed? Can't you use the `esbmc parseoptions` directly?

> If we create a new flag, would that be a duplicate?

I am still confused about what this does for this frontend. For the C frontend, this flag is used to define the entry point (as a `main`).

kunjsong01 replied:
For the Solidity frontend, this flag is used in the same way as the C frontend - it also defines the entry point. 

> Are you sure this is needed? 

Yes, it will be eventually used as the target_func in `src/solidity-frontend/pattern_check.h`.

> Can't you use the `esbmc parseoptions` directly?

If I remembered correctly, there were a bunch of patches to migrate to `boost::program_options`, which makes `parseoptions` obsolete.  (cf. [cmdline.cpp history: commits on May 22, 2021](https://github.com/esbmc/esbmc/commits/master/src/util/cmdline.cpp) ). I used `vm.count("function")` to reuse the '--function' flag for Solidity frontend.

rafaelsamenezes replied:
> If I remembered correctly, there were a bunch of patches to migrate to boost::program_options, which makes parseoptions obsolete

I mean [esbmc_parseoptions](https://github.com/esbmc/esbmc/blob/master/src/esbmc/esbmc_parseoptions.cpp), which sets the [options object](https://github.com/esbmc/esbmc/blob/master/src/esbmc/options.cpp).

kunjsong01 replied:
Thanks very much for the clarification! 

Given the fact that we want to re-use the '--function' flag in exactly the same way as C frontend, a new flag would become duplicate. There is no conflict to the existing C frontend because it does not break the C verification flow.  

For the time being, '--function' is used verify all functions in a Solidity smart contract one by one. I'm not 100% sure but there might be a more elegant way to verify a Solidity smart contract as a whole, and hence this line is probably open to further changes. I'll put a comment in the next patch set.

mikhailramalho replied:
I agree with @rafaelsamenezes that this seems odd: to make it consistent with how the other frontends work `--function` should only check one function, while when not using `--function` all functions would be checked.

kunjsong01 replied:
Sorry, my previous comment might have caused some confusion about `--function` flag in the new Solidity frontend. 

For Solidity forntend, the `--function`flag only checks one function, which is consistent with other frontends. E.g. 
```
esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_overflow --compact-trace
```

> For the time being, '--function' is used verify all functions in a Solidity smart contract one by one.

What I really mean by this comment is that the above command can be wrapped in an automation script to verify all functions one by one,  something like: 
```
esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_1 --compact-trace
esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_2 --compact-trace
esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_3 --compact-trace
```

Apologies for the misleading comment. I was coding an automation framework to run more complex Solidity smart contract that contains multiple functions, while I was writing the previous comment. These thoughts accidently interfered and slipped into the writing. 😄

rafaelsamenezes replied:
Ok, so you shouldn't need to recreate this flag in `cmdline`, you can use from `options`.

Now I am confused about the C modules though. Could you document the full expected flow with some examples (or regression cases)?

kunjsong01 replied:
> Could you document the full expected flow with some examples (or regression cases)?

Sure, I'll attach a short PDF to outline the big picture as well as a few examples.

mikhailramalho commented:
> Most of the environment already works through vcpkg (at least in windows). The exceptions are the solvers since only Z3 is available in vcpkg
> But can't you also use cmake's fetch thing to build a dependency? I thought I read this somewhere. @kunjsong01, this discussion is not related to your PR, just the part to add the support for the json library via vcpkg.
> […](#)
> -- Mikhail Ramalho.

Any update on removing the json library from the commit? If you don't know how to do it, you can use tools like BFG.

kunjsong01 commented:
> > Most of the environment already works through vcpkg (at least in windows). The exceptions are the solvers since only Z3 is available in vcpkg
> > But can't you also use cmake's fetch thing to build a dependency? I thought I read this somewhere. @kunjsong01, this discussion is not related to your PR, just the part to add the support for the json library via vcpkg.
> > […](#)
> > -- Mikhail Ramalho.
> 
> Any update on removing the json library from the commit? If you don't know how to do it, you can use tools like BFG.

WIP. I'm going to use cmake's FetchContent. 

BTW, what is BFG?

mikhailramalho commented:
You need to remove all the json library code you pushed from the commit
history.

BFG is tool that can help you clean your git history, you can google it.

Em qua., 15 de set. de 2021 às 10:47, KSong ***@***.***>
escreveu:

> Most of the environment already works through vcpkg (at least in windows).
> The exceptions are the solvers since only Z3 is available in vcpkg
> But can't you also use cmake's fetch thing to build a dependency? I
> thought I read this somewhere. @kunjsong01 <https://github.com/kunjsong01>,
> this discussion is not related to your PR, just the part to add the support
> for the json library via vcpkg.
> … <#m_-7574758029077905504_>
> -- Mikhail Ramalho.
>
> Any update on removing the json library from the commit? If you don't know
> how to do it, you can use tools like BFG.
>
> WIP. I'm going to use cmake's FetchContent.
>
> BTW, what is BFG?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/493#issuecomment-920088026>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH33IIH3KIBZJBTYBJLUCCWZFANCNFSM5D574E7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

mikhailramalho commented:
I'm still a bit confused about this but I'll try to understand what you
mean when I'll star to test the patch.

On Wed, 15 Sep 2021, 11:58 KSong, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/util/cmdline.cpp
> <https://github.com/esbmc/esbmc/pull/493#discussion_r709329218>:
>
> > @@ -121,6 +123,14 @@ bool cmdlinet::parse(
>
>      args = vm["input-file"].as<std::vector<std::string>>();
>
>      verification_file = args.back();
>
>    }
>
> +  if (vm.count("dummymain")) // dummy main for Solidity frontend
>
> +  {
>
> +    sol_main = vm["dummymain"].as<std::vector<std::string>>().back();
>
> +  }
>
> +  if (vm.count("function")) // We need Solidity function name for pattern-based verification
>
> +  {
>
> +    sol_function = vm["function"].as<std::string>();
>
>
> Sorry, my previous comment might have caused some confusion about
> --function flag in the new Solidity frontend.
>
> For Solidity forntend, the --functionflag only checks one function, which
> is consistent with other frontends. E.g.
>
> esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_overflow --compact-trace
>
>
> For the time being, '--function' is used verify all functions in a
> Solidity smart contract one by one.
>
> What I really mean by this comment is that the above command can be
> wrapped in an automation script to verify all functions one by one,
> something like:
>
> esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_1 --compact-trace
>
> esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_2 --compact-trace
>
> esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast -M main.c --function func_3 --compact-trace
>
>
> Apologies for the misleading comment. I was coding an automation framework
> to run more complex Solidity smart contract that contains multiple
> functions, while I was writing the previous comment. These thoughts
> accidently interfered and slipped into the writing. 😄
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/493#discussion_r709329218>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZCP2RVXAXMS4HE2CLUCC7CDANCNFSM5D574E7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kunjsong01 committed:
Moved json library to ExternalDependencies

kunjsong01 committed:
Added solidity frontend as a option

kunjsong01 commented:
Hi guys, 

Thanks a lot for the review comments. 

Following the comments, I've done two more patches that: 

1. Removed JSON library. Added it as an external dependency in `scripts/cmake/ExternalDependencies.cmake` 
2. Added the new Solidity frontend as an opton 'ENABLE_OLD_FRONTEND' in `scripts/cmake/Options.cmake`. It won't get compiled unless the user gives `-DENABLE_SOLIDITY_FRONTEND=On` - pretty much like the old frontend.

The code base looks more clean and tidy after removing the huge blob of JSON stuff. However, I'm not sure how we to proceed with this PR. Here are my thougths and questions: 

Q1. Shall I create another branch in my forked repository, something like "dev-solidity-merge" to amalgate all commits as a single commit? This PR contains 97 commits. If it is merged in the future, it will "contaminate" the commit history of ESBMC main. If anything is horribly broken due to this new frontend, it would be easier to revert one commit. 

Q2. It appears that I probably have fiddled with the main branch in my [forked ESBMC repository](https://github.com/kunjsong01/esbmc/commits/master?after=c34a9f968e229fabb71dfd572dce720dc0cc80dd+34&branch=master). I've got a pair of duplicate commits: 

![image](https://user-images.githubusercontent.com/32592800/133608823-1f8ab394-bae4-4dfb-9243-04f635d53981.png)

To make matters worse, these duplicate commits appeared in this PR: 

![image](https://user-images.githubusercontent.com/32592800/133609207-a915a604-cb35-4bc1-a575-149deafef470.png)

Any suggestions to fix this problem? My conjecture is that it cannot be fixed because it's not possible to "re-write" the commit history.

mikhailramalho commented:
> Hi guys,
> 
> Thanks a lot for the review comments.
> 
> Following the comments, I've done two more patches that:
> 
> 1. Removed JSON library. Added it as an external dependency in `scripts/cmake/ExternalDependencies.cmake`

@kunjsong01, you've added a new commit removing the old json lib. As I already said twice, **this needs to be removed from the git hsitory**. You can use a tool like BFG to do this.

> 2. Added the new Solidity frontend as an opton 'ENABLE_OLD_FRONTEND' in `scripts/cmake/Options.cmake`. It won't get compiled unless the user gives `-DENABLE_SOLIDITY_FRONTEND=On` - pretty much like the old frontend.
> 
> The code base looks more clean and tidy after removing the huge blob of JSON stuff. However, I'm not sure how we to proceed with this PR. Here are my thougths and questions:
> 
> Q1. Shall I create another branch in my forked repository, something like "dev-solidity-merge" to amalgate all commits as a single commit? This PR contains 97 commits. If it is merged in the future, it will "contaminate" the commit history of ESBMC main. If anything is horribly broken due to this new frontend, it would be easier to revert one commit.

Please, don't squash the commits. It's easier to have a sequence of commits than one massive one. It also make it easier to revert individual commits.

> 
> Q2. It appears that I probably have fiddled with the main branch in my [forked ESBMC repository](https://github.com/kunjsong01/esbmc/commits/master?after=c34a9f968e229fabb71dfd572dce720dc0cc80dd+34&branch=master). I've got a pair of duplicate commits:
> 
> ![image](https://user-images.githubusercontent.com/32592800/133608823-1f8ab394-bae4-4dfb-9243-04f635d53981.png)
> 
> To make matters worse, these duplicate commits appeared in this PR:
> 
> ![image](https://user-images.githubusercontent.com/32592800/133609207-a915a604-cb35-4bc1-a575-149deafef470.png)
> 
> Any suggestions to fix this problem? My conjecture is that it cannot be fixed because it's not possible to "re-write" the commit history.

You did a few merges there. You need to rebase your commits on top of master's HEAD.

kunjsong01 commented:
@mikhailramalho Hi Mikhail, thanks very much for the replies. They are very helpful. 

In order to have a clean commit history, I'm planning to do the following things: 
 
1. Use BFG to trim away the json lib files in all commits. 
2. Sync my forked master and rebase the feature branch (new frontend) to get rid of the duplicate commits that would otherwise contaminate ESBMC mainline history.

kunjsong01 commented:
Abandoned.

Superseded by ESBMC Solidity support #497 .
